[JENKINS-28676] Serve artifacts using the default encoding of the JVM

rkivisto opened the pull request:
With this change, archived artifacts will be served with the `contenttype` header set to the default encoding of the JVM.

One example where this is useful is if the `file.encoding` system property is `UTF-8`, then this change will update the HTML `contenttype` header to have `charset=utf-8`, so that archived artifacts with unicode content are rendered correctly in the browser.

Ajay Kumar suggested this improvement over in the comments of [JENKINS-28676](https://issues.jenkins.io/browse/JENKINS-28676?focusedCommentId=387265&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-387265)



### Testing done

I tested this fails with Jenkins LTS 2.361.2 manually by:

1. Starting a fresh instance of Jenkins LTS 2.361.2, with only the 'Recommended plugins' installed, with startup option `-Dfile.encoding=UTF-8`
2. Adding this Pipeline job config.xml on disk at JENKINS_HOME/jobs/test/config.xml and restarting the instance:[config.xml.zip](https://github.com/jenkinsci/jenkins/files/9871024/confix.xml.zip)
3. Running the above job.
4. Accessing build 1, and clicking on the archived artifact `Test.log`, it will not render the unicode content correctly:
![Jenkins-2 361 2](https://user-images.githubusercontent.com/19917557/198071109-146a0510-c9f3-40ad-9177-7030bb6fe281.gif)

I tested that my changes fix this issue by doing the same steps as above with my Jenkins core:
![wip-fix](https://user-images.githubusercontent.com/19917557/198072244-772e8dc7-c5a3-4ca7-b9a3-1720402d8a3f.gif)

I also tried to write an automated testcase for this, which I was able to get the negative assertions working (to confirm it fails when `file.encoding` is not set to `UTF-8`) , but my positive assertions (to confirm my fix is working when `file.encoding` is set to `UTF-8`) does not work:

```
            //TODO not working
            //assertTrue(contentType.contains(utf8Header));
            //String pageContent=page.getWebResponse().getContentAsString();
            //assertTrue("unicode not rendered with file.encoding set to UTF-8", pageContent.contains("✅"));
```

After review I have removed this testcase from the PR.

I manually tested with the following startup argument combinations and then inspecting the `Content-Type` response header. For example when running with `-Dfile.encoding=UTF-8` the header was `text/plain;charset=utf-8`:
<img width="606" alt="UTF-8" src="https://user-images.githubusercontent.com/19917557/200438511-8244d964-645d-4538-82a8-6ed68988cfbf.png">

Here are the startup argument combinations I manually tested:

| Startup argument | Content-Type response header |
| --- | ---- |
| -Dfile.encoding=UTF-8 | text/plain;charset=utf-8 |
| -Dfile.encoding=ISO-8859-1 | text/plain;charset=iso-8859-1 |
| -Dfile.encoding=Windows-1252 | text/plain;charset=Windows-1252 |
| -Dfile.encoding=SHIFT-JIS | text/plain;charset=SHIFT-JIS |
| -Dfile.encoding=UTF-8 -Dhudson.model.DirectoryBrowserSupport.disableArtifactDefaultCharset=true | text/plain |

I will open a separate PR to document the escape hatch to disable my code: https://github.com/jenkins-infra/jenkins.io/blob/master/content/doc/book/managing/system-properties.adoc

### Proposed changelog entries

- Serve archived artifacts using the default encoding of the JVM

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] The Jira issue, if it exists, is well-described.
- [x] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)).
  - Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [x] There is automated testing or an explanation as to why this change has no tests.
- ~[] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.~
- ~[] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.~
- ~[ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).~
- ~[ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.~
- ~[ ] For new APIs and extension points, there is a link to at least one consumer.~

### Desired reviewers


### Maintainer checklist

Before the changes are marked as `ready-for-merge`:

- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).


<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/7308"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>

rkivisto committed:
set charset to UTF-8 when serving a file

rkivisto committed:
JENKINS-28676 serve unicode artifacts when file.encoding has UTF-8 in it

rkivisto committed:
JENKINS-28676 serve unicode artifacts when "-Dfile.encoding" is "UTF-8"

rkivisto committed:
JENKINS-28676 partially working testcase

NotMyFault committed:
Merge branch 'master' into JENKINS-28676

basil reviewed (commented):
In practice the default character set specified by `file.encoding` is almost always UTF-8 except when running on Windows, and even there it will be UTF-8 starting in Java 18 via [JEP 400](https://openjdk.org/jeps/400). I think the main question here is whether we want to assume that build artifacts are in the character set returned by `java.nio.charsets.Charset.defaultCharset()` (prior to Java 18, usually UTF-8 on Unix and something else on Windows, and starting with Java 18 usually UTF-8 everywhere) or in UTF-8 specifically. If the latter, we could just hard-code UTF-8 in the header without worrying about the value of `file.encoding`. But I think the former is preferable to give users the most flexibility. In that case, though, why only add `charset=UTF-8` to the header when `java.nio.charsets.Charset.defaultCharset()` is UTF-8 (or, equivalently, `file.encoding` is UTF-8)? If the underlying logic is that artifacts should be served in the JVM's default encoding, then it seems we should set `charset` to whatever the JVM's default encoding is, regardless of whether or not that happens to be UTF-8.

rkivisto commented:
I was purposely leaving this fix very narrow in scope to try to limit causing regressions, and did not want to try setting the charset for every instance since I was not able to get my testcase for this fully working.

rkivisto committed:
[JENKINS-28676] fix checkstyle warnings

basil commented:
Please generalize the solution to allow for any supported JVM character encoding, not just UTF-8. There is no need to be afraid of regressions as long as you test with a few common character encodings, such as ISO-8859-1, Windows-1252, and perhaps something like Shift JIS. You will not be able to write an automated test for this, but you should be able to test the change manually (and update the **Testing done** section with the results) by changing the JVM's default encoding and verifying that the appropriate `charset` value is set in the HTTP response for each encoding.

rkivisto committed:
[JENKINS-28676] serve artifacts using the default encoding of the JVM

rkivisto committed:
[JENKINS-28676] remove test since it is not working

rkivisto commented:
Thanks for the advice @basil! I've now made the solution generic for any JVM encoding, removed the automated testcase that was incomplete, and also updated the `Testing done` section with the combination of startup arguments I tested manually, and what the `Content-Type` was in the response header for that test.

rkivisto committed:
Merge branch 'jenkinsci:master' into JENKINS-28676

daniel-beck commented:
I don't understand this PR at all.

Jenkins doesn't know the encoding of files it serves. Given how common it is for such files to not be generated by Jenkins core or agent processes (most archived artifacts), even the `-Dfile.encoding` doesn't look like a reasonable approximation here. What am I missing?

daniel-beck commented on core/src/main/java/hudson/model/DirectoryBrowserSupport.java:
```diff
@@ -375,6 +378,14 @@ private void serveFile(StaplerRequest req, StaplerResponse rsp, VirtualFile root
+            String fileEncoding = System.getProperty("file.encoding");
+            if (fileEncoding != null) {
+                // JENKINS-28676 serve artifacts using the default encoding of the JVM
+                rsp.setContentType("text/plain;charset=\"" + fileEncoding + "\"");
```
This makes the comment in line 401 obsolete.

daniel-beck replied:
Also, looks like https://github.com/jenkinsci/stapler/blob/3aa95fe179cb5534806b05d2f41947791a78b928/core/src/main/java/org/kohsuke/stapler/Stapler.java#L662-L663 via https://github.com/jenkinsci/stapler/blob/3aa95fe179cb5534806b05d2f41947791a78b928/core/src/main/java/org/kohsuke/stapler/Stapler.java#L571-L572 makes https://github.com/jenkinsci/stapler/blob/3aa95fe179cb5534806b05d2f41947791a78b928/core/src/main/java/org/kohsuke/stapler/Stapler.java#L132-L140 the canonical way to do this.

rkivisto commented:
Thanks for the review @daniel-beck I think this needs to be done in a different way, I'll close this one for now.
